[REVIEW] Initial support for auto-ARIMA

Nyrio opened the pull request:
## Introduction

Auto-ARIMA is a popular feature available in R with the `forecast` and `fable` packages, in Python with the `pmdarima` module (a wrapper around `statsmodels`), etc. The purpose of Auto-ARIMA is to select a good ARIMA model automatically, with a process first described in ["Automatic Time Series Forecasting: The `forecast` Package for R", Rob J. Hyndman & Yeasmin Khandakar (2008)](https://doi.org/10.18637/jss.v027.i03).

This PR brings a batched support with a friendly Python interface and great performance. I'll go through some of the details below.

## The interface

The user has a high level of control through the many options supported by the `search` method (which provides good default values to get started easily too). Here is an example of using auto-ARIMA to forecast seasonal time series:

```python
model = AutoARIMA(y)
model.search(s=12, d=(0, 1), D=(0, 1), p=(0, 2, 4), q=(0, 2, 4),
             P=range(2), Q=range(2), method="css", truncate=100)
model.fit(method="css-ml")
fc = model.forecast(20)
```

When calling `search`, the user specifies the model space by providing an integer/sequence/generator for each hyper-parameter (or relying on the default values). They can also specify which fitting method to use during the search (see below), which stationarity and seasonality tests to use, etc. The user can then call `fit` and `predict`/`forecast` like a normal `ARIMA` model, even though multiple models are used behind the scenes (see below).

## An abstraction around multiple `ARIMA` models

Different series in the batch might not work well with the same models, so this class will divide the batch into a bunch of pieces (_sub-batches_) and associate each series to one of the pieces. It works by building two internal arrays to keep track of which piece each series belongs to, and the position of the series in this sub-batch. After computing predictions separately, it puts the results back together so that the user doesn't have to deal with this complexity.

To really take advantage of the GPU acceleration, we need batch_size >> n_models, which should be the case if the model space is a few dozens and the batch size in the thousands.

## The sum-of-squares method for performance

As we need to fit a lot of models in parallel, the step-wise search method used in other implementations wouldn't be efficient for us. At the moment `search` is using a grid search for a few parameters that are determined by minimizing an information criterion. In order to perform this search as fast as possible, I have implemented an alternative method to fit a model, which is using a simple sum-of-squares instead of the complex statespace methods. This can give very decent estimations of the parameters for the temporary models, while being orders of magnitude faster. I have included a few optimizations in ARIMA to accelerate this method even further but the bottleneck is now the Python L-BFGS-B for this method.

## Stationarity and seasonality test

For the stationarity test, I have refactored a bit the existing KPSS test, and made the implementation modular to be able to use other stationarity tests later if we add them.

At the moment the seasonality test is based on the STL decomposition from statsmodels, and could be a bottleneck for large batches. I plan to add an implementation of the STL decomposition in cuML later to fix that.

## TODO-list

- [x] P0: proper logging/tracking
- [ ] P1: C++/CUDA STL decomposition
- [ ] P2: Box-Cox transformations
- [ ] P2: C++/CUDA L-BFGS-B (@rietmann-nv)
- [ ] P2: enable other stationarity/seasonality tests
- [ ] ??: one-fits-all method?

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

Nyrio commented on python/cuml/tsa/stationarity.pyx:
```diff
@@ -18,115 +18,96 @@
+        input_to_cuml_array(y, check_dtype=[np.float32, np.float64])
+    cdef uintptr_t d_y_ptr = d_y.ptr
+
+    if output_type == "input":
```
`_input_to_type` doesn't take into account the specified `output_type`, so if the user wants a specific output type that is different from the input type we shouldn't override it. Which is why it is called only when the user wants the output type to mirror the input type.

Or am I missing something?

Nyrio commented:
@dantegd I have updated my branch and opened the issues #2232, #2233, #2234 and #2235. I have left one discussion open above.

I have also fixed the `test_fit_function` test for ARIMA. It is still failing for `AutoARIMA` because I am not sure exactly whether I should change the interface or skip the test. Basically it's a made up model inspired from a feature available in R and pmdarima. The interface I have done uses 2 functions `search` and `fit` instead of simply `fit`. The reason is that you may want to do a search, check what models have been selected, maybe do it again etc, and then at the end compute the expensive `fit` operation. Does that interface sound reasonable? Should I update/skip the test?

KeertiBafna commented:
@Nyrio  Will we be able to use auto_ARIMA in the next release ?

Nyrio commented:
@KeertiBafna It won't be part of 0.14 but this PR should be merged in early 0.15, and improvements should follow in separate PRs

Nyrio commented:
@dantegd I've added `AutoARIMA` to the xfail list in `test_fit_function` and incorporated the changes that I PRed separately in #2340 to fix the test for `ARIMA` too. It should hopefully pass CI now.

dantegd commented:
rerun tests

Nyrio commented:
rerun tests

teju85 commented:
@dantegd can we get an approval from python side, please?

Nyrio commented:
rerun tests

KeertiBafna commented:
@Nyrio I can see auto_arima in the 0.15 nightly version. Can we start using it for some initial testing ?

Nyrio commented:
@KeertiBafna It is indeed available in the nightly 0.15, and I would welcome feedback about the interface, features, etc. Please be aware that this initial support still has some performance bottlenecks that will be addressed in the future. Feel free to tag me in issues related to ARIMA and AutoARIMA.

Upcoming features in ARIMA (such as support for prediction intervals, missing observations, etc) will be added in AutoARIMA at the same time.

KeertiBafna commented:
@Nyrio Thank for you the quick support. I have been indeed waiting for this feature. 
I have tested it on a batch of my data, and here a few questions:
1) Can we not allow d(non-seasonal differencing to exceed 1 ? 
I get " TypeError: 'int' object is not iterable " for the below code:
#code
from cuml.tsa.auto_arima import AutoARIMA
est = AutoARIMA(grp_df)
# Search zone
t0 = time.time()
est.search(s=52, 
           d=list(range(2)), ## 
           D=list(range(2)), 
           p=list(range(4)), 
           q=list(range(4)), 
           P=list(range(2)), 
           Q=list(range(2)), 
           )
If i use a list(range(1)) for d, the problem solves and i get my forecasts. 

2) Currently i can only get a summary of order and seasonal order combinations. Can we not get the order and seasonal order chosen per ts in the batch ? This would really help if i want to re-run things with additional data but not want to search again.

Nyrio commented:
@KeertiBafna 

1. I cannot reproduce the exception with the same parameters as your example. At the moment d can be 0, 1 or 2, as long as d+D <= 2, so your code should be valid. Could you please provide a few more details? Batch size, number of observations, and the exception details?

2. Thanks for this feedback. Feel free to file a feature request issue, and suggest how you would like this feature to work (e.g do we provide a slice, do we keep an option for the short summary, etc?)

KeertiBafna commented:
@Nyrio 

Batch size: 10 
Number of Observations: 173
Exception:
""
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-9c194a0b5470> in <module>
      9            q=range(4),
     10            P=range(2),
---> 11            Q=range(2))

cuml/tsa/auto_arima.pyx in cuml.tsa.auto_arima.AutoARIMA.search()

TypeError: 'int' object is not iterable
""

Code block:
from cuml.tsa.auto_arima import AutoARIMA
est = AutoARIMA(grp_df)
t0 = time.time()
est.search(s=52, 
           d=range(2), 
           D=range(2), 
           p=range(4), 
           q=range(4), 
           P=range(2), 
           Q=range(2))

Nyrio commented:
@KeertiBafna I opened #2651 to fix this bug, sorry for the delay.
